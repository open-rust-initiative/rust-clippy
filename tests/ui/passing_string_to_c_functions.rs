#![allow(unused)]
#![warn(clippy::passing_string_to_c_functions)]
#![feature(rustc_private)]
extern crate libc;

use std::ffi::{c_char, CString};
use libc::strlen;

extern "C" {
    fn greet(name: *const c_char);
}

fn some_string() -> String {
    "hello".to_string()
}

fn greet_safe(name: *const c_char) {}

fn main() {
    let name = String::from("Rust");
    let cstr_name = CString::new("Rust").unwrap();
    unsafe {
        greet(name.as_ptr() as *const _); // lint
        greet(cstr_name.as_ptr() as *const _); // don't lint
        greet(some_string().as_ptr() as *const _); // FN, we currently don't lint calls
        strlen(name.as_ptr() as *const _); // lint
        let c_str_ptr = cstr_name.as_ptr() as *const c_char;
        strlen(c_str_ptr); // don't lint
    }

    greet_safe(name.as_ptr() as *const _); // don't lint
}
