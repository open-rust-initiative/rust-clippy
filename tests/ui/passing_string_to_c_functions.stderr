error: passing native strings to external functions
  --> $DIR/passing_string_to_c_functions.rs:23:9
   |
LL |         greet(name.as_ptr() as *const _); // lint
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: use `CString` or `CStr` instead
  --> $DIR/passing_string_to_c_functions.rs:23:15
   |
LL |         greet(name.as_ptr() as *const _); // lint
   |               ^^^^
   = note: `-D clippy::passing-string-to-c-functions` implied by `-D warnings`

error: passing native strings to external functions
  --> $DIR/passing_string_to_c_functions.rs:26:9
   |
LL |         strlen(name.as_ptr() as *const _); // lint
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: use `CString` or `CStr` instead
  --> $DIR/passing_string_to_c_functions.rs:26:16
   |
LL |         strlen(name.as_ptr() as *const _); // lint
   |                ^^^^

error: aborting due to 2 previous errors

